### ISSUE TITLE
Non-equilibrium plasma including EEDF computation (two-term approximation)

### ISSUE THREAD

**Issue by qumale** [Cantera/enhancements#194]
---
name: Non-equilibrium plasma including EEDF computation (two-term approximation)
about: Computation of an EEDF providing a set of electron collision cross sections and a reduced electric field E/N
title: Non-equilibrium plasma including EEDF
labels: feature-request
assignees: '@QuentinMale, @NicolasBarleon'

---

**Abstract**

We aim to solve 
$\frac{\partial}{\partial \varepsilon}\left(\tilde{W} F_0-\tilde{D} \frac{\partial F_0}{\partial \varepsilon}\right)=\tilde{S}+\tilde{R}$
using numerical methods described in Ref. [1] to obtain rate coefficients for the plasma reactions
$k_k = \gamma \int_0^\infty \varepsilon \sigma_k F_0 d \varepsilon$ .

**Motivation**

Describe the need for the proposed change:

* What problem is it trying to solve? 
Solving EEDF to define reaction rate for electron collision reactions.
Reproducing the work of Ref. [2] for a complete non-equilibrium plasma modeling.
* Who is affected by the change?
Users of the plasma WIP feature.
* Why is this a good solution?
We need to solve the EEDF to be able to compute the plasma reaction rates. The approach is robust and has been used and validated in Refs [1,2].

**Possible Solutions**

Implementation of the numerical methods of Ref. [1] as already done in the open-source code BOLOS (https://github.com/aluque/bolos). Use of the work of @BangShiuh for the cross section reading and usage.

Implementation of new reaction types may be needed to take into account feeding of vibrational energy during collisional processes and fast gas heating (see Section 3.2 Ref. [2]).

**Work in progress branch(es)**
https://github.com/QuentinMale/canteraBS/tree/collision-reaction
https://github.com/QuentinMale/canteraBS/tree/canteraPlasma

**References**

[1] Hagelaar, G. J. M., & Pitchford, L. C. (2005). Solving the Boltzmann equation to obtain electron transport coefficients and rate coefficients for fluid models. Plasma Sources Science and Technology, 14(4), 722–733. https://doi.org/10.1088/0963-0252/14/4/011
[2] Cheng, L., Barleon, N., Cuenot, B., Vermorel, O., & Bourdon, A. (2022). Plasma assisted combustion of methane-air mixtures: Validation and reduction. Combustion and Flame, 240, 111990. https://doi.org/10.1016/j.combustflame.2022.111990

**Comment by BangShiuh** [comment]
The plasma reaction rates (electron collision reactions) is determined by EEDF which is a phase property like temperature. Two-term approximation needs E/N (electric field strength divided by gas number density) to calculate EEDF. Electric field is an external force but not a phase property. I would suggest making a class PlasmaReactor with a member variable electricFieldStrength. We will put PlasmaPhase object into PlasmaReactor, and the EEDF can be calculated in the reactor and set for PlasmaPhase.

**Comment by QuentinMale** [comment]
We have a working prototype here: [https://github.com/QuentinMale/canteraBS/tree/collision-reaction](https://github.com/QuentinMale/canteraBS/tree/collision-reaction)

See commits and samples/python/thermo/plasmatest.py

This is a first prototype for demonstration of feasibility. Will require clean-up, and development for proper interface between PlasmaPhase and the EEDF solver in EEDFTwoTermApproximation. Before continuing the work, I suggest we have a chat so that everyone agrees on the design.

**Comment by BangShiuh** [comment]
https://github.com/Cantera/cantera/pull/1262#issuecomment-2002203452 could be considered.

**Comment by qumale** [comment]
We have recently made a number of new developments, including the creation of a "PlasmaReactor" reactor class containing additional equations to take account of vibrational energy, and the modification of the RHS term in the temperature equation to take account of discharge heating.

https://github.com/QuentinMale/canteraBS/tree/canteraPlasma

**Comment by BangShiuh** [comment]
Hi All,

https://github.com/BangShiuh/cantera/tree/collision-new This is the branch
I just wrote to make PlasmaPhase using Kinetics and
ElectronCollisionPlasmaRate. This implementation uses weak_ptr,
registerChangeCallback, and store ElectronCollisionPlasmaRate as
m_collisionRates. I hope this implementation satisfies the requirements and
can be useful for future development. I will take some time to finish
addressing the issues for the PR
https://github.com/Cantera/cantera/pull/1731 this month.

Best,
Bang

**Comment by Gaetanosaure** [comment]
The last developments made on the PlasmaReactor from Nicolas Barléon, Quentin Malé and Bang-Shiuh Chen done by Gaétan Bizot (PhD student of Antonio Andreini and of Nicolas Barléon) currently present the features described below. It can be found: [here](https://github.com/QuentinMale/canteraBS/tree/canteraPlasma )

1) Implementation of the vibrational energy equation inside of the solver. This allows to collapse all the equations of the vibrational levels of a given species into one equation.
This implementation is available for as many species as the user wishes.

2) It is fully controlled via the yaml file. Species having their vibrational levels represented by a vibrational energy equation must be declared in the phase with the key vib_species. For example: ‘vib_species = [N2, O2, CH4, H2O]’. For electron-collision reactions that involve those species, one should add the key ‘vib_bool: true’. The code retrieves itself the energy threshold of the reactions and the target species. 

3) There is also the possibility for any other reaction’s energy to be split between vibrational energy and usual energy release. This is specified by a set of three parameters: “vib_bool: true, vib_target: N2, d_u_vib: 2.19” for example.

4) This vibrational energy can be relaxed in three ways. The way of computing the relaxation time is chosen by the user via the python interface. The three options are the following: a constant relaxation time, a relaxation time based on the article [‘Modelling the impact of non-equilibrium discharges on reactive mixtures for simulations of plasma-assisted ignition in turbulent flows’](https://doi.org/10.1016/j.combustflame.2016.01.009) of Castela et al. from 2016 and relaxation time based on the article [‘Plasma-assisted ignition and combustion’ ](http://dx.doi.org/10.1016/j.pecs.2012.05.003) of Starikovskiy and Aleksandrov from 2013. For the latter development a .yaml file is included within the source code. It holds the coefficients used by the formula computing the relaxation time of a species into another. It is based on the article of Starikovskiy and Aleksandrov. This data file is used by default, but the experimented user thinking that the coefficients should be modified can provide a .yaml file containing customized coefficients for vibrational species relaxation time via the pythoninterface without the need to recompile everything.

5) Development of a constant pressure version of the plasma reactor. Both are of interest since a discharge is initially constant volume and rapidly evolves towards constant pressure.

**Comment by ischoegl** [comment]
@Gaetanosaure ... thank you for the update! 

~Would you be willing to convert this into a pull request so it can be merged with the main Cantera repository? The reason I am asking is that once a contribution is merged, it will be automatically maintained for the foreseeable future.~

_Edit: nevermind, this is Cantera/cantera#1887_

**Comment by NicolasBarleon** [comment]
Hi @ischoegl ,

Thank you for linking this with the other pull request. We’re currently awaiting input from @mquiram and @speth to confirm whether they are also working on integrating our plasma reactor development. This will help us avoid duplicating efforts. Once clarified, we’d be very happy to merge all these developments into an official Cantera release!

Since you're more familiar with the overall development, do you think it would be better to continue the work here or in the other pull request? We're happy to proceed in whichever way you think is most appropriate.

**Comment by BangShiuh** [comment]
@NicolasBarleon I think it is better to break the project into several short-term achievable PRs. For example, the vibrational energy equation is a great idea, but it requires a significant work. This can be a separate PR as we should be able to develop the vibrational energy equation without two-term approximation feature. I will be happy to share my experience on merging code into Cantera on a call.

**Comment by NicolasBarleon** [comment]
@BangShiuh, we've been progressing with the development in an incremental fashion, following three main steps: starting with the EEDF implementation, then moving on to the plasma reactor without vibrational effects, and finally adding the vibrational energy equations.

We felt it was challenging to develop all these components in parallel, especially in terms of validation—we’ve been comparing results against our previous code to ensure consistency. One issue is that without the EEDF equation, the energy transfer from electrons to the gas isn't well-defined. While we could potentially reformulate it, we're not sure it would bring much added value. You can see the results we've obtained so far in @mquiram’s pull request, which reflect our current implementation.

_but it requires a significant work_
I think it's amost finished!

We can send you a poll by email for the next week to organize a call, lets see if we find a slot?

**Comment by ischoegl** [comment]
Thanks, @NicolasBarleon. I'm not quite certain about the constellation, so I agree that some coordination is warranted. Generally speaking, I agree with @BangShiuh to slice efforts into smaller pieces, but I also realize that this is not always possible.

**Comment by speth** [comment]
I've updated Cantera/cantera#1887 to clarify the authorship of various parts of this implementation, at least as far as I understand it based on the history of the [canteraPlasma](https://github.com/QuentinMale/canteraBS/tree/canteraPlasma branch) in @QuentinMale's fork. @NicolasBarleon and @BangShiuh, I'd appreciate your confirmation that this is at least reasonably correct.

I'd like to use this Enhancement thread to discuss some potential updates to the current PR (I find that "issues" are better than PR threads at not hiding important conversations and being polluted by information about different commits that get pushed).

The first of these how we provide/store information about electron collision processes. Currently in the PR, there are two ways of doing this. The first, already in use in the `main` branch, uses the `electron-collision-plasma` rate type and the corresponding `ElectronCollsionPlasmaRate` class, with a YAML representation like:
```yaml
- equation: CO2 + e => e + e + CO2+
  type: electron-collision-plasma
  energy-levels: [13.7, 14.0, 14.5, ..., 800.0, 900.0, 1000.0, 9000.0]
  cross-sections: [0.0, 2e-22, 5.6e-22, ..., 1.47e-20, 1.4e-20, 0.0]
```
The second is an entry in a `collisions` list which is represented as an `ElectronCrossSection` object in C++, and has a representation like:
```yaml
- target: N2
  product: N2(rot)
  data:
  - [0.02, 0.0]
  - [0.03, 2.5e-22]
  ...   
  - [3.6, 1.7e-21]
  - [5.0, 0.0]
  threshold: 0.02
  kind: excitation
```
There are only a few meaningful differences between these formats, and I think it would be beneficial if we could avoid introducing this second format and storage class. The additional data fields (`threshold`, `kind`) can be added to the existing reaction type with no issue. The only challenge is converting the `target` and `product` into a reaction `equation`, given the variability in what is provided in different sources of cross section data. In this example, `N2(rot)` is not define as a species in the phase, so this equation would not currently satisfy the checks that Cantera usually makes on these equations and might require a bit of a workaround. Other than that, my suggestion would be to write the second collision as:
```yaml
- equation: e + N2 => e + N2(rot)
  type: electron-collision-plasma
  kind: excitation
  threshold: 0.02
  energy-levels: [0.02, 0.03, ... 3.6, 5.0]
  cross-sections: [0.0, 2.5e-22, ..., 1.7e-21, 0.0]
```
I had discussed some time ago with @BangShiuh that collisions of this form, which are irrelevant from the standpoint of the gas phase kinetics, could be listed in a separate "collisions" section, following the same format as the normal `reactions` section.

**Comment by BangShiuh** [comment]
The unification of collision cross sections and collision reactions is a good approach. I would like to clarify the usage. Depending on the intended use of the data, we could organize it in the input yams file as follows:
	•	If the data is used only for solving the electron Boltzmann equation, it should be placed under the field collisions.
	•	If the data is used only for calculating reaction rate constants, it should be placed under the reactions section.

•	If the data is used for both reaction rate constant calculations and solving the Boltzmann equation, it should be placed under the reactions section, but with an additional field: solve-boltzmann: true. (Field name to be decided)

Does this sound reasonable? Hope this clarify the input format.

(The authorship is fine with me.)

**Comment by NicolasBarleon** [comment]
I agree that the definition of the cross-section needs to be clarified, and even unified

What I’m certain about is that, if an EEDF solver is used, it should be placed within the collision section. The rationale is that **solving the EEDF is only meaningful when based on a consistent and validated set of cross-sections**. Including this in the reaction section could be confusing and prone to errors.

For reaction rates without an EEDF solver, are you considering using an analytical distribution like a Maxwellian? In that case, it would make sense to include the corresponding rates directly in the reaction section. However, **it might be clearer to define a unique cross-section format in the collision section**, from which specific reactions can be referenced by matching targets and products. This would promote consistency and simplify the structure, similar to how it’s handled in the LXCat database and bolsig+ in ZDPlasKin. Additionnaly to simplify the implementation of analytical EEDF, it could be just an option replacing the **converge** method in the EEDF solver by a prescribed EEDF.

Regarding the reaction e + N2 => e + N2(rot) It’s no longer necessary to explicitly include this in the mechanism as e + N2 → e + N2, since we now account for plasma power through a source term in the energy equation. This reaction was previously used as a workaround to include the heating effect in the absence of a power source term. The current formulation is more consistent and robust and no longer requires such approximations/tricks.

However, for the mean vibrational energy, this kind of workaround is still needed. For example, the vibrational excitation reaction e + N2 → e + N2 should still be represented as e + N2 → e + N2 with the associated energy threshold accounted for in the vibrational energy equation, and excluded from the gas energy equation (Ep - Evib). We didn't encounter issue by using this kind of reaction in the branch associated to this enhancement. Is it something new in Cantera to could not allow working like that with reactions?

I hope that I provided clarification about these two points. Do not hesitate if you need more details. :)

As for the history issue, @speth, I’ll take a look after a few days off.

**Comment by NicolasBarleon** [comment]
@speth Git history is fine to us too.

What is the current status of Matthew's PR? Is there anything missing, or is it ready to merge?

Also, did you have a chance to think about the **PlasmaReactor** feature, including the additional vibrational energy equations and plasma source terms we introduced in the related issue?

**Comment by speth** [comment]
Resolved by completion of Cantera/cantera#1887.
